### ISSUE TITLE
cuda graphs dont play nice with numpy numbers

### ISSUE THREAD

**Issue by ben-da6** [pytorch/pytorch#172942]
### 🐛 Describe the bug

In the below example the use of np.sqrt (which will produce a np.float64) , causes a cuda graph partition. We have instances in our code where this sort of thing actually breaks the compile, ultimately segfaulting. Can this be handled better.

```python
import numpy as np
import torch


class MyMod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(10, 10)
        self._float = np.sqrt(4.0)

    def forward(self, x):
        return self.linear(x) * self._float


if __name__ == "__main__":
    model = MyMod()
    model.to("cuda")
    model.compile(mode="max-autotune", dynamic=False)
    example_input = torch.randn(1, 10, device="cuda")

    torch.compiler.reset()
    with torch._inductor.utils.fresh_inductor_cache():
        model(example_input)

```

### Error logs

```
/persist/envs/Fluyt312/lib/python3.12/site-packages/torch/_inductor/compile_fx.py:312: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
  warnings.warn(
cudagraph partition due to non gpu ops
cudagraph partition due to DeviceCopy ops
cudagraph partition into 2 partitions
```

### Versions

torch==2.9.1

cc @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng @chauhang @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260518+cu130`) on NVIDIA H100 and confirmed it reproduces (3/3 runs failed).

All 3 runs: torch.compile emitted a 'DeviceCopy in input program' warning when compiling MyMod (which stores np.sqrt(4.0) as an attribute) with mode='max-autotune', dynamic=False. Exit code 1 on all 3 runs. Detection check: 'DeviceCopy in input program' matched True; 'cudagraph partition into ...' and 'cudagraph partition due to ...' matched False in this nightly.

<details>
<summary>Error output</summary>

```
Backend: inductor, PyTorch: 2.13.0.dev20260518+cu130

--- Captured output ---
/home/karthickps/.conda/envs/pytorch-nightly/lib/python3.13/site-packages/torch/_inductor/compile_fx.py:320: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
  warnings.warn(
W0519 01:53:32.726000 825334 site-packages/torch/_inductor/utils.py:2970] [0/0_1] DeviceCopy in input program

--- end ---

'cudagraph partition into ...':   False
'cudagraph partition due to ...': False
'DeviceCopy in input program':    True

test_fn equivalent: False
Bug PRESENT (cudagraph partition / device copy detected)

```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import sys
import tempfile
import numpy as np
import torch


class MyMod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(10, 10)
        self._float = np.sqrt(4.0)

    def forward(self, x):
        return self.linear(x) * self._float


def _capture_fd_output(fn):
    """Run fn() while capturing both stdout and stderr at the FD level
    (so Python logging which writes to fd=2 via cached handlers is also captured)."""
    sys.stdout.flush()
    sys.stderr.flush()
    with tempfile.TemporaryFile(mode="w+") as tmp_out, tempfile.TemporaryFile(mode="w+") as tmp_err:
        saved_out = os.dup(1)
        saved_err = os.dup(2)
        os.dup2(tmp_out.fileno(), 1)
        os.dup2(tmp_err.fileno(), 2)
        try:
            fn()
        finally:
            sys.stdout.flush()
            sys.stderr.flush()
            os.dup2(saved_out, 1)
            os.dup2(saved_err, 2)
            os.close(saved_out)
            os.close(saved_err)
        tmp_out.seek(0)
        tmp_err.seek(0)
        return tmp_out.read() + tmp_err.read()


def _compile_and_run():
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")
    model = MyMod()
    model.to("cuda")
    example_input = torch.randn(1, 10, device="cuda")
    with torch._inductor.utils.fresh_inductor_cache():
        model.compile(mode="max-autotune", dynamic=False, backend=backend)
        out = model(example_input)
    return model, out


def _count_partitions(model) -> int:
    """Try to introspect the compiled model for partition count.
    Returns the number of partitions if discoverable, else -1."""
    try:
        runner = None
        for attr in ("_compiled_call_impl", "_orig_mod"):
            if hasattr(model, attr):
                pass
        for obj in vars(model).values():
            if hasattr(obj, "partitions"):
                return len(obj.partitions)
        return -1
    except Exception:
        return -1


def test_fn() -> bool:
    """Return True if NO bug (no cudagraph partition). False if bug present."""
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")

    captured = _capture_fd_output(_compile_and_run)

    if backend != "inductor":
        return True

    lower = captured.lower()
    partition_into = "cudagraph partition into" in lower
    partition_due_to = "cudagraph partition due to" in lower
    devicecopy_in_input = "devicecopy in input program" in lower
    return not (partition_into or partition_due_to or devicecopy_in_input)


if __name__ == "__main__":
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")
    print(f"Backend: {backend}, PyTorch: {torch.__version__}")

    torch._dynamo.reset()
    try:
        captured = _capture_fd_output(_compile_and_run)
    except Exception:
        import traceback; traceback.print_exc()
        os._exit(2)

    print("\n--- Captured output ---")
    print(captured if captured else "(empty)")
    print("--- end ---\n")

    lower = captured.lower()
    partition_into = "cudagraph partition into" in lower
    partition_due_to = "cudagraph partition due to" in lower
    devicecopy_in_input = "devicecopy in input program" in lower

    print(f"'cudagraph partition into ...':   {partition_into}")
    print(f"'cudagraph partition due to ...': {partition_due_to}")
    print(f"'DeviceCopy in input program':    {devicecopy_in_input}")

    if backend != "inductor":
        ok = True
    else:
        ok = not (partition_into or partition_due_to or devicecopy_in_input)

    print(f"\ntest_fn equivalent: {ok}")
    print(f"Bug {'NOT present' if ok else 'PRESENT (cudagraph partition / device copy detected)'}")
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260518+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] onnx==1.21.0
[pip3] onnx-ir==0.2.1
[pip3] onnxscript==0.7.0
[pip3] torch==2.13.0.dev20260517+cu130
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260517+cu130
[pip3] torchdata==0.11.0
[pip3] torchinfo==1.8.0
[pip3] torchmetrics==1.9.0
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260518+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.
